### DISCUSSION TITLE
How do people use pond?

### DISCUSSION THREAD

**Discussion by PatrickF1** [marcransome/pond#90] [Show and tell]
I understand the idea of pond and I think the implementation and interface are beautiful. I just don't see how it would be useful for me. What are some real-world problems that pond is helping to solve?

**Comment by jaminthorns** [comment]
I was just using Pond earlier today and wanted to thank @marcransome for his continued maintenance and polish of this niche plugin. I saw this discussion and figured it would be a good place to share some thoughts. I also [submitted a PR to awsm.fish](https://github.com/jorgebucaran/awsm.fish/pull/138) to let more people know about it.

I've [used Pond for a couple years now](https://github.com/jaminthorns/environment/commit/76d8e7c28164da7c162ad426577c768348034dfb), and it's been really nice for my setup. I have a multi-machine development environment where I use a Git repository to keep in sync across computers. Sometimes, though, I'll have configuration that I only want on a specific computer, usually my work laptop. I've got a really basic way to handle per-machine differences with file templating, but that's not really appropriate for multiple machine-specific files.

As I've developed on the same application for several years in my current job, I've written several helper functions in Fish that make some common tasks really easy, but these tasks are really specific to that application and wouldn't make sense to persist alongside my general development environment. I _could_ put these functions directly in my `~/.config/fish/functions` directory, but then they'd be scattered among my general use functions that I sync with the aforementioned Git repo. A natural solution would be to store these functions in some other directory and add that directory to `fish_function_path` so they get autoloaded by Fish. If I wanted to organize these functions into groups, then it would make sense to add multiple directories.

Another bit of configuration I like to do is to set up variables for quick access to common directories (especially if they differ across machines), like `$DOWNLOADS`, `$SCREENSHOTS`, `$GOOGLE_DRIVE`, etc. I could add those to my `config.fish` or a `conf.d` file, but that has the same issue of mixing machine-specific config among my general use config.

It turns out Pond suits these 2 use cases perfectly, and along with it provides a really nice CLI and grouping of functions/variables ("ponds"). It works quite nicely for me, and it's nice that I don't have to manually create directories and append to `fish_function_path`. I use it mainly as a tool for organization, not a necessity but a nicety.

I don't use the enable/disable functionality (I just keep everything enabled), but I imagine it could be really useful to emulate [`direnv`](https://direnv.net)-like functionality by enabling/disabling a pond in a handler function using `--on-variable PWD`.

**Reply by marcransome** [reply]
@jaminthorns It's been enlightening to read how other people are finding use for `pond` besides myself. I'd be interested to hear any thoughts you might have on the future direction of the project. Thanks for the praise. 👍🏻

**Reply by PatrickF1** [reply]
Thank you for explaining @jaminthorns! That makes sense, I don't have the same problems to solve as you so that's probably why I never thought of that!

**Comment by marcransome** [comment]
@PatrickF1 Apologies for the ludicrously slow response. Your message must have passed me by somehow. 😳

The original idea for `pond` spawned from a desire to have a mechanism for organising functions into distinct groups, partly due to the large number of functions that I maintain across multiple systems. Most of those have similar—typically short—names, which introduces some ergonomic annoyance when attempting to tab-complete function names when everything is _always_ accessible (i.e. when all functions are housed in `~/.config/fish/functions`) so `pond` helps with that.

Besides that, I frequently interact with a number of varied environments (VMs, containers, remote and cloud-based environments) and have built up distinct sets of functions over the years for each of those that I can make accessible (or not) with `pond` at any given time. I use the `autoload`/`autounload` functionality to manipulate the shell prompt to add context to what functions are available at any given time (in a manner similar to [starship](https://github.com/starship/starship)).

As an offshoot of `pond`, I setup [fish-shop](https://github.com/fish-shop) as a place to house several `fish`-related custom GitHub actions, a number of which are used in workflows in this project.

**Reply by marcransome** [reply]
I should add, [fzf](https://github.com/PatrickF1/fzf.fish) is one of my most used `fish` plugins, so a _thank you_ is certainly in order. 🥳

**Reply by PatrickF1** [reply]
You're welcome and thanks for commenting to tell me :)
